File: frontend/src/types/experiments.ts
export interface DesignFactor {
  name: string
  levels: unknown[]
  // Short, user-editable names for individual treatments. They are persisted
  // alongside raw levels and become categorical factor values in the Results
  // CSV, so a long system prompt never becomes CSV data outside its secure
  // execution record.
  level_labels?: string[]
  // Absent means 'string' -- factors created before this field existed keep
  // working unchanged. Drives which control the factor editor renders per
  // level (see components/protocol/factorLevels.ts); purely a frontend/UX
  // concern, not enforced by the backend (design_spec is opaque JSONB).
  level_type?:
    | 'string'
    | 'text'
    | 'number'
    | 'boolean'
    | 'llm_config'
    | 'tool_config'
    | 'pattern'
    | 'script_config'
    | 'dataset_config'
    | 'tool_names'
}

export interface DesignMetric {
  // Stable instance identity.  Catalog entries also retain their catalog key;
  // custom and legacy declarations intentionally have no catalog key.
  id?: string
  catalogKey?: string
  name: string
  description?: string
  kind?: 'runtime' | 'custom'
  valueType?: 'number' | 'boolean' | 'string' | 'opaque'
  unit?: string
  primary: boolean
  direction: 'maximize' | 'minimize' | 'neutral'
  aggregation?: 'mean' | 'sum' | 'none'
  [key: string]: unknown
}

export type MeasurementProducerKind = 'runtime' | 'reported'
export type MeasurementDirection = 'maximize' | 'minimize' | 'neutral'
export type MeasurementAggregation = 'mean' | 'sum' | 'rate' | 'pooled' | 'none'

export interface MeasurementPlan {
  metrics: Array<{
    id: string
    name: string
    value_type: 'number' | 'boolean' | 'opaque'
    direction: MeasurementDirection
    aggregation: MeasurementAggregation
    primary: boolean
    description?: string
    unit?: string
  }>
  producers: Array<{
    id: string
    producer_id: string
    kind: MeasurementProducerKind
    outputs: Record<string, string>
    artifacts: string[]
    config: Record<string, unknown>
  }>
  inputs: Array<{
    producer_binding_id: string
    input_key: string
    source_key: string
  }>
}

export interface MeasurementPlanValidationReport {
  valid: boolean
  issues: Array<{ code: string; message: string; path: string; blocking?: boolean }>
}

export interface MeasurementCapabilities {
  outputs: Record<string, string[]>
}

export type ObservationStatus =
  | 'measured'
  | 'unavailable'
  | 'failed'
  | 'timed_out'
  | 'cancelled'
  | 'not_applicable'

// Every slug here is implemented -- there is deliberately no "coming soon"
// entry. Six ARES coordination-category placeholders (supervisor, swarm, task
// bidding, supervision tree, event-driven, multi-agent planning) used to be
// listed and were removed: an option that always fails at run time is worse
// than an option that isn't offered. The backend still recognizes the five that
// remain unbuilt, so an experiment saved with one gets a real explanation, not
// "unknown" -- supervisor came back off that list once it was actually built.
export type CoordinationStrategySlug =
  | 'sequential'
  | 'critic_gate'
  | 'peer_collaboration'
  | 'supervisor_architecture'

export interface CoordinationStrategyConfig {
  slug: CoordinationStrategySlug
  params?: Record<string, unknown>
}

export const COORDINATION_STRATEGY_CATALOG: {
  slug: CoordinationStrategySlug
  label: string
  description: string
}[] = [
  {
    slug: 'sequential',
    label: 'Sequential (default)',
    description: "Each agent's output becomes the next agent's input, following the canvas's own edges in order.",
  },
  {
    slug: 'critic_gate',
    label: 'Critic Gate',
    description: 'A reviewer agent approves or requests revisions at a fixed point in the sequential pipeline.',
  },
  {
    slug: 'peer_collaboration',
    // "(beta)" rides on the label rather than a separate badge for the same
    // reason "(default)" does above: the label is what the picker, the
    // canvas-mismatch warning and the switch-confirm dialog all render, so
    // one string keeps the qualifier from appearing in only one of the three.
    label: 'Peer Collaboration (beta)',
    // The cost ceiling belongs here rather than in RunConfirmDialog: this is
    // where the choice is actually made, and every consultation is a full agent
    // run of its own (Reason/Plan/Act cycle, own tokens). The caps are the
    // messenger's -- services/agent_messenger.py.
    description:
      'Connected agents work the task together as a conversation -- the lead agent can consult its peers, and every reply is shared with everyone, instead of handing off once. Each consultation is a full agent run (up to 8, nested 2 deep), so a cell can cost several times a sequential one.',
  },
  {
    slug: 'supervisor_architecture',
    label: 'Supervisor (beta)',
    // The contrast with Peer Collaboration is the whole reason to pick one over
    // the other, so it's stated rather than left to be discovered at run time:
    // there, the lead *may* consult; here, ASAREE dispatches every worker
    // itself. The turn count is the topology's, which is why no cap is quoted.
    description:
      'One supervisor agent briefs the workers wired to it, they all run (in parallel by default), an optional reviewer assesses their output advisorily, and the supervisor writes the final answer. Every one of those turns is dispatched by ASAREE, so no worker can be skipped -- a cell costs one agent run per agent, plus a second for the supervisor.',
  },
]


export interface DesignSpec {
  factors?: DesignFactor[]
  // Copies per factor-level combination (default 1 when absent).
  replicates?: number
  // Shuffles generated cells' execution order when set (never affects which
  // combinations/replicates are generated).
  randomization_seed?: number | null
  // Declared up front, unlike the Results tab's purely-inferred metric keys
  // -- lets the Design tab show "Metrics" before any cell has run.
  metrics?: DesignMetric[]
  coordination_strategy?: CoordinationStrategyConfig
  // Which staged pipeline this experiment's dataset workspaces use. There is
  // deliberately no UI for this: a canvas-built experiment leaves it absent and
  // the backend derives the plan from which stage-writing MCP servers the
  // canvas wires. It stays declared here (untyped -- the shape is
  // asaree_workspace_core's, not the GUI's) purely so a spread that rebuilds
  // design_spec carries an SDK-declared plan through instead of erasing it.
  stage_plan?: unknown
  [key: string]: unknown
}

export interface Experiment {
  id: string
  name: string
  description: string | null
  hypothesis: string | null
  design_type: string
  task_brief: Record<string, unknown> | null
  design_spec: DesignSpec | null
  measurement_plan: MeasurementPlan | null
  metric_recommendations?: MetricRecommendationMetadata | null
  metric_recommendation_set_version?: number
  latest_test_run_id?: string | null
  // Every dataset attached to this experiment, in canvas wiring order -- an
  // experiment can run against several since the Dataset connector was
  // uncapped. `dataset_id` is a read-only view of the first one, kept for
  // code written before that; it is no longer a stored column.
  dataset_ids: string[]
  dataset_id: string | null
  locked_at: string | null
  locked_protocol_revision_id: string | null
  locked_design_spec: DesignSpec | null
  locked_measurement_plan: MeasurementPlan | null
  created_at: string
  updated_at: string
  archived_at: string | null
}

export interface MetricRecommendationMetadata {
  applied_version: number | null
  dismissed_version: number | null
  intentionally_removed_keys: string[]
  contextual_suggestion_dismissals?: Record<string, string>
}

// One row of the Runs tab's trial list -- one replicate, not ProtocolRun; a replicate that's
// never been run at all is still a trial, reported with status "not_started".
// Matches src/asaree/api/experiments.py's TrialResponse exactly.
export interface Trial {
  replicate_label: string
  factor_values: Record<string, unknown>
  metric_values: Record<string, unknown>
  status: 'not_started' | 'queued' | 'running' | 'finalizing' | 'completed' | 'failed' | 'cancelled'
  run_id: string | null
  // True when this run used an older published canvas version than the
  // protocol's current published version.
  obsolete: boolean
  error: string | null
  updated_at: string
}

// One row of analysis["emm_cells"] -- one factor-level combination's
// estimated marginal mean, with its own CI (the Results tab's "uncertainty"
// ask). _condition_label matches services.design_generation.cell_label_for's
// own format (e.g. "tier_large"), computed fresh from condition_factors, not
// stored anywhere.
export interface EmmCell {
  _condition_label: string
  mean: number
  std: number
  count: number
  se: number
  ci_lo: number
  ci_hi: number
}

// One row of analysis["factorial_effects"] -- "effect" has no ":" for a
// single factor's main effect, one or more ":"-joined factor names for an
// interaction (see services.factorial_analysis._design_matrix's own
// ":".join(...) term naming).
export interface FactorialEffect {
  effect: string
  estimate_half_diff: number
  t: number
  p_perm: number
  p_maxstat_fwer: number
  mc_se_p: number
}

export interface NonInferiorityRow {
  condition: string
  contrast_vs_reference: number
  lower_bound: number
  neg_delta: number
  p_one_sided: number
  p_holm?: number
  ni_decision?: string
  [key: string]: unknown
}

// The full services.factorial_analysis.analyze_factorial return shape --
// deliberately loose (an index signature, not every key modeled) since this
// is a dict[str, Any] on the backend too, not a typed Pydantic model.
export interface ExperimentAnalysis {
  n_attempted: number
  n_scored: number
  n_failed: number
  n_not_yet_run: number
  emm_cells: EmmCell[]
  factorial_effects: FactorialEffect[]
  non_inferiority: NonInferiorityRow[]
  ni_reportable: boolean
  metric_summary: Record<string, unknown>[]
  cost_time_summary: Record<string, unknown>[]
  footer: { primary_metric: string; condition_factors: string[]; [key: string]: unknown }
  [key: string]: unknown
}

// GET /experiments/{id}/results -- available is false (with a human-
// readable reason) whenever there's nothing to show yet: no factors/primary
// metric declared, a factor with other than 2 levels, or not enough scored
// replicates (see services.factorial_analysis.analyze_experiment_design).
export interface ExperimentResults {
  available: boolean
  reason: string | null
  analysis: ExperimentAnalysis | null
  best_condition: EmmCell | null
}

// The general-purpose Results panel is intentionally not limited to a
// balanced factorial design. These records combine execution facts (status,
// duration, usage) with whatever numeric metrics the experiment produced.
export interface ResultNodeRun {
  node_id: string
  node_label: string
  status: string
  output_text: string | null
  error: string | null
  // Senders this node's prompt referenced that produced nothing, by label
  // (already resolved server-side -- see experiment_run_results). Empty in the
  // normal case. The assembled prompt just has a gap where their output should
  // be, so without this it reads as an agent that was never told anything.
  unresolved_reference_labels?: string[]
  agent_run_id: string | null
  input_tokens: number | null
  output_tokens: number | null
  total_tokens: number | null
  cost_usd: number | null
}

export interface MetricObservation {
  metric_id: string
  metric_name: string
  value_type: 'number' | 'boolean' | 'opaque'
  status: ObservationStatus
  value: unknown
  error: string | null
  attempt_id: string
  producer: {
    binding_id: string
    producer_id: string
    kind: MeasurementProducerKind | 'unbound'
    version: string
    binding_config?: Record<string, unknown>
  }
  input_provenance: Record<string, { source_key: string; value_type: string; provenance: Record<string, unknown> }>
}

export interface EvaluationArtifact {
  artifact_key: string
  kind: string
  payload: unknown
  attempt_id: string
  producer: MetricObservation['producer']
  input_provenance: MetricObservation['input_provenance']
}

export interface LegacyMetricValue {
  metric_id: string
  metric_name: string
  value: unknown
  attempt_id: string
  producer: {
    binding_id: 'legacy-unknown'
    producer_id: 'legacy.unknown'
    kind: 'legacy'
    version: 'unknown'
  }
}

export interface ResultReplicate {
  replicate_label: string
  replicate_number: number
  cell_label: string
  factor_values: Record<string, unknown>
  metric_values: Record<string, unknown>
  status: Trial['status']
  obsolete: boolean
  // The prior attempt is obsolete; this stable replicate slot has no result
  // yet against the current canvas version.
  requires_current_attempt?: boolean
  error: string | null
  run_id: string | null
  protocol_revision_id: string | null
  updated_at: string
  duration_seconds: number | null
  node_runs: ResultNodeRun[]
  input_tokens: number | null
  output_tokens: number | null
  total_tokens: number | null
  cost_usd: number | null
  agent_run_count: number
  reported_usage_count: number
  reported_cost_count: number
  metric_evaluation?: {
    status: 'queued' | 'running' | 'completed' | 'failed' | 'skipped'
    error?: string | null
    evaluator_run_id?: string | null
    metric_ids?: string[]
  } | null
  metric_observations: MetricObservation[]
  evaluation_artifacts: EvaluationArtifact[]
  legacy_values?: LegacyMetricValue[]
  obsolete_runs: ObsoleteRun[]
  superseded_runs: SupersededRun[]
}

// Read-only facts for a past ProtocolRun attempt. Both obsolete and
// superseded attempts stay available for inspection, but never participate in
// current results, CSV, or statistical analysis.
export interface HistoricalRun {
  run_id: string
  status: Trial['status']
  obsolete: boolean
  error: string | null
  protocol_revision_id: string | null
  updated_at: string
  metric_values: Record<string, unknown>
  metric_evaluation?: ResultReplicate['metric_evaluation']
  metric_observations: MetricObservation[]
  evaluation_artifacts: EvaluationArtifact[]
  legacy_values?: LegacyMetricValue[]
  duration_seconds: number | null
  node_runs: ResultNodeRun[]
  input_tokens: number | null
  output_tokens: number | null
  total_tokens: number | null
  cost_usd: number | null
  agent_run_count: number
  reported_usage_count: number
  reported_cost_count: number
}

// A prior attempt used an older published canvas revision.
export interface ObsoleteRun extends HistoricalRun {
  obsolete: true
}

// A prior attempt used the current canvas revision, but was replaced by a
// later run of the same stable replicate slot.
export interface SupersededRun extends HistoricalRun {
  obsolete: false
}

export interface ResultCell {
  cell_label: string
  factor_values: Record<string, unknown>
  replicate_count: number
  completed_count: number
  current_completed_count: number
  obsolete_count: number
  metric_means: Record<string, number>
  metric_counts: Record<string, number>
  cost_usd: number | null
  total_tokens: number | null
  duration_seconds: number | null
}

export interface RunResultsOverview {
  total_replicates: number
  completed_replicates: number
  running_replicates: number
  queued_replicates: number
  failed_replicates: number
  not_started_replicates: number
  obsolete_replicates: number
  total_cost_usd: number | null
  total_input_tokens: number | null
  total_output_tokens: number | null
  total_tokens: number | null
  total_duration_seconds: number | null
  agent_run_count: number
  reported_usage_count: number
  reported_cost_count: number
}

export interface ExperimentRunResults {
  overview: RunResultsOverview
  metric_keys: string[]
  metric_types: Record<string, 'number' | 'boolean'>
  metric_aggregations: Record<string, 'mean' | 'sum'>
  metric_directions: Record<string, MeasurementDirection>
  primary_metric: string | null
  primary_metric_direction: 'maximize' | 'minimize' | null
  cells: ResultCell[]
  replicates: ResultReplicate[]
}

export interface Replicate {
  // One independently runnable observation within the owning cell.
  id: string
  cell_id: string
  cell_label: string
  replicate_label: string
  replicate_number: number
  // Which generation of the design this observation was made under. Replicates
  // from a superseded revision stay in the database as history, so listReplicates
  // returns only the current revision's unless asked for another one.
  design_revision_id: string
  run_id: string | null
  workspace_id: string | null
  factor_values: Record<string, unknown> | null
  metric_values: Record<string, unknown> | null
  artifacts: Record<string, unknown> | null
  created_at: string
  updated_at: string
}

// One generation of an experiment's design. Regenerating a design whose set
// of cells has changed supersedes the current revision and opens a new one,
// keeping the old cells (and anything scored in them) as history rather than
// deleting them -- `superseded_at === null` marks the one that is current.
export interface DesignRevision {
  id: string
  revision: number
  superseded_at: string | null
  design_spec: DesignSpec | null
  cell_count: number
  replicate_count: number
  scored_replicate_count: number
  created_at: string
}

export interface DesignImpact {
  has_generated_design: boolean
  regeneration_required: boolean
  current_cell_count: number
  proposed_cell_count: number
  added_cell_count: number
  retained_cell_count: number
  removed_cell_count: number
  current_replicate_count: number
  proposed_replicate_count: number
  added_replicate_count: number
  retained_replicate_count: number
  removed_replicate_count: number
  // Why an update is needed, not just that it is. 'coordination_strategy_changed'
  // is the reason the counts above can't express -- it adds and removes no
  // cells, so on its own it reads as "no change" beside the banner.
  regeneration_reasons: (
    | 'no_design_generated'
    | 'coordination_strategy_changed'
    | 'design_matrix_changed'
    | 'cells_drifted'
  )[]
}

File: sdk/src/asaree_client/resources/experiments.py
"""Experiment and factorial-cell resource, matching asaree.api.experiments.

``upsert_cell`` is the replacement for the source notebook's two
``client.runs.update(mlm_run_id, metadata={...})`` calls (pre-scoring and
post-scoring) — both become calls to this, merged onto the same
``FactorialCellResult`` row rather than the run.
"""

from __future__ import annotations

import builtins
import uuid
from datetime import datetime
from typing import Any

from asaree_client.models import Cell, DesignRevision, Experiment, ExperimentArtifact

ResourceId = uuid.UUID | str
# Distinguishes "omit this kwarg" (leave unchanged) from "pass None"
# (explicitly clear/detach) in update() below -- a plain default of None
# can't tell those apart.
_UNSET: Any = object()


class Experiments:
    def __init__(self, client: Any) -> None:
        self._client = client

    def create(
        self,
        *,
        name: str,
        description: str | None = None,
        design_type: str = "factorial",
        task_brief: dict[str, Any] | None = None,
        factors: builtins.list[dict[str, Any]] | None = None,
        dataset_ids: builtins.list[ResourceId] | None = None,
    ) -> Experiment:
        payload: dict[str, Any] = {"name": name, "design_type": design_type}
        if description is not None:
            payload["description"] = description
        if task_brief is not None:
            payload["task_brief"] = task_brief
        if factors is not None:
            payload["factors"] = factors
        if dataset_ids is not None:
            payload["dataset_ids"] = [str(d) for d in dataset_ids]
        data = self._client._post("/experiments", json=payload)
        return Experiment(**data)

    def get(self, experiment_id: ResourceId) -> Experiment:
        data = self._client._get(f"/experiments/{experiment_id}")
        return Experiment(**data)

    def list(self) -> builtins.list[Experiment]:
        data = self._client._get("/experiments")
        return [Experiment(**e) for e in data]

    def delete(self, experiment_id: ResourceId) -> None:
        self._client._delete(f"/experiments/{experiment_id}")

    def update(
        self,
        experiment_id: ResourceId,
        *,
        name: str | None = _UNSET,
        description: str | None = _UNSET,
        hypothesis: str | None = _UNSET,
        dataset_ids: builtins.list[ResourceId] | None = _UNSET,
        dataset_id: ResourceId | None = _UNSET,
        design_spec: dict[str, Any] | None = _UNSET,
        archived_at: datetime | None = _UNSET,
    ) -> Experiment:
        """Only the fields actually passed are sent (omit one to leave it
        unchanged; pass ``None`` explicitly to clear/detach it) --
        ``dataset_ids`` replaces the whole attached list (``[]`` detaches
        everything) and ``dataset_id`` is the one-dataset shorthand, with
        ``dataset_id=None`` attaching nothing / detaching, same as before;
        ``name``/``description``/``hypothesis`` are for editing an
        experiment's own identity/write-up straight from the SDK (matching
        the GUI's Rename/Edit description/Design-tab hypothesis fields).
        ``design_spec`` is a full replacement, not a merge -- the same
        factors/replicates/randomization_seed/metrics/coordination_strategy
        dict the Design tab renders; read the current value with ``get()``
        first if you only want to change one key. ``archived_at`` archives
        (pass a datetime) or unarchives (pass ``None`` explicitly) -- the
        GUI's own Archive/Unarchive action, and the only way end users are
        meant to remove an experiment from their active list (the GUI no
        longer exposes ``delete()`` at all, to prevent accidental data
        loss; it's still here for scripted cleanup)."""
        payload: dict[str, Any] = {}
        if name is not _UNSET:
            payload["name"] = name
        if description is not _UNSET:
            payload["description"] = description
        if hypothesis is not _UNSET:
            payload["hypothesis"] = hypothesis
        if dataset_ids is not _UNSET:
            payload["dataset_ids"] = [str(d) for d in dataset_ids] if dataset_ids else []
        if dataset_id is not _UNSET:
            payload["dataset_id"] = str(dataset_id) if dataset_id else None
        if design_spec is not _UNSET:
            payload["design_spec"] = design_spec
        if archived_at is not _UNSET:
            payload["archived_at"] = archived_at.isoformat() if archived_at else None
        data = self._client._patch(f"/experiments/{experiment_id}", json=payload)
        return Experiment(**data)

    def generate_design(self, experiment_id: ResourceId) -> builtins.list[Cell]:
        """Materialize one cell per combination of the experiment's declared
        factors, returning the current design's cells.

        Safe to call again after changing the factors. If the new design
        produces a different set of cells than the current one, the current
        design revision is superseded and a new one opened: results for cells
        that survive the change carry forward, and the rest stay behind in the
        superseded revision as history (see ``list_design_revisions``). Nothing
        is deleted, and the returned list is always exactly the new design.
        """
        data = self._client._post(f"/experiments/{experiment_id}/generate-design")
        return [Cell(**c) for c in data]

    def list_design_revisions(self, experiment_id: ResourceId) -> builtins.list[DesignRevision]:
        """Every generation of this experiment's design, newest first. The
        first entry (``superseded_at is None``) is the current design."""
        data = self._client._get(f"/experiments/{experiment_id}/design-revisions")
        return [DesignRevision(**r) for r in data]

    def delete_design_revision(self, experiment_id: ResourceId, revision_id: ResourceId) -> None:
        """Permanently delete a superseded design revision and all of its
        cells, including any results scored in them. Refuses (409) on the
        current revision -- regenerate the design to replace that instead."""
        self._client._delete(f"/experiments/{experiment_id}/design-revisions/{revision_id}")

    def upsert_cell(
        self,
        experiment_id: ResourceId,
        cell_label: str,
        *,
        run_id: ResourceId | None = None,
        workspace_id: str | None = None,
        factor_values: dict[str, Any] | None = None,
        metric_values: dict[str, Any] | None = None,
        artifacts: dict[str, Any] | None = None,
    ) -> Cell:
        """Merge fields onto a cell's row — pass just what changed; unset
        fields are left untouched (a pre-scoring call and a post-scoring
        call land on the same row without either erasing the other)."""
        payload: dict[str, Any] = {}
        if run_id is not None:
            payload["run_id"] = str(run_id)
        if workspace_id is not None:
            payload["workspace_id"] = workspace_id
        if factor_values is not None:
            payload["factor_values"] = factor_values
        if metric_values is not None:
            payload["metric_values"] = metric_values
        if artifacts is not None:
            payload["artifacts"] = artifacts
        data = self._client._put(f"/experiments/{experiment_id}/cells/{cell_label}", json=payload)
        return Cell(**data)

    def get_cell(self, experiment_id: ResourceId, cell_label: str) -> Cell:
        data = self._client._get(f"/experiments/{experiment_id}/cells/{cell_label}")
        return Cell(**data)

    def list_cells(self, experiment_id: ResourceId, *, revision_id: ResourceId | None = None) -> builtins.list[Cell]:
        """The experiment's current design cells, or -- with *revision_id* --
        a superseded revision's, to read results scored under an older
        design."""
        params = {"revision_id": str(revision_id)} if revision_id is not None else None
        data = self._client._get(f"/experiments/{experiment_id}/cells", params=params)
        return [Cell(**c) for c in data]

    def analyze(
        self,
        experiment_id: ResourceId,
        *,
        condition_factors: builtins.list[str],
        positive_levels: dict[str, Any],
        reference_condition: dict[str, Any],
        primary_metric: str,
        alpha: float = 0.05,
        delta: float = 0.05,
        n_resamples: int = 10_000,
        seed: int = 42,
        failure_flag_key: str = "failure_flag",
        cost_keys: builtins.list[str] | None = None,
    ) -> dict[str, Any]:
        """Run the spinal_surgery use case's specific statistical methodology
        (Freedman-Lane + max-stat FWER, BCa non-inferiority + Holm) against
        this experiment's current cells. Not the generic nonparametric-
        regression capability tracked separately (ASAREE#1)."""
        payload: dict[str, Any] = {
            "condition_factors": condition_factors,
            "positive_levels": positive_levels,
            "reference_condition": reference_condition,
            "primary_metric": primary_metric,
            "alpha": alpha,
            "delta": delta,
            "n_resamples": n_resamples,
            "seed": seed,
            "failure_flag_key": failure_flag_key,
        }
        if cost_keys is not None:
            payload["cost_keys"] = cost_keys
        return self._client._post(f"/experiments/{experiment_id}/analyze", json=payload)  # type: ignore[no-any-return]

    def create_artifact(
        self, experiment_id: ResourceId, *, name: str, kind: str, content: str
    ) -> ExperimentArtifact:
        """A durable landing spot for anything worth keeping past one run --
        e.g. ``analyze(...)``'s own result (``kind="analyze_result"``,
        ``content=json.dumps(result)``) or a flattened CSV export
        (``kind="csv_export"``), in place of writing either to local disk.
        Create-once/append-style: calling this again with the same ``name``
        creates a NEW row, it never overwrites the last one."""
        data = self._client._post(
            f"/experiments/{experiment_id}/artifacts", json={"name": name, "kind": kind, "content": content}
        )
        return ExperimentArtifact(**data)

    def list_artifacts(self, experiment_id: ResourceId) -> builtins.list[ExperimentArtifact]:
        data = self._client._get(f"/experiments/{experiment_id}/artifacts")
        return [ExperimentArtifact(**a) for a in data]

    def get_artifact(self, experiment_id: ResourceId, artifact_id: ResourceId) -> ExperimentArtifact:
        data = self._client._get(f"/experiments/{experiment_id}/artifacts/{artifact_id}")
        return ExperimentArtifact(**data)

    def delete_artifact(self, experiment_id: ResourceId, artifact_id: ResourceId) -> None:
        self._client._delete(f"/experiments/{experiment_id}/artifacts/{artifact_id}")
